File: src/com/trollworks/gcs/character/IdentityPanel.java
/*
 * Copyright (c) 1998-2016 by Richard A. Wilkes. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public License,
 * version 2.0. If a copy of the MPL was not distributed with this file, You
 * can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This Source Code Form is "Incompatible With Secondary Licenses", as defined
 * by the Mozilla Public License, version 2.0.
 */

package com.trollworks.gcs.character;

import com.trollworks.gcs.page.DropPanel;
import com.trollworks.toolkit.annotation.Localize;
import com.trollworks.toolkit.ui.layout.ColumnLayout;
import com.trollworks.toolkit.utility.Localization;

import javax.swing.SwingConstants;

/** The character identity panel. */
public class IdentityPanel extends DropPanel {
	@Localize("Identity")
	@Localize(locale = "de", value = "Identität")
	@Localize(locale = "ru", value = "Личность")
	@Localize(locale = "es", value = "Identidad")
	private static String	IDENTITY;
	@Localize("Name:")
	@Localize(locale = "de", value = "Name:")
	@Localize(locale = "ru", value = "Имя:")
	@Localize(locale = "es", value = "Nombre:")
	private static String	NAME;
	@Localize("Title:")
	@Localize(locale = "de", value = "Titel:")
	@Localize(locale = "ru", value = "Статус:")
	@Localize(locale = "es", value = "Título:")
	private static String	TITLE;
	@Localize("Religion:")
	@Localize(locale = "de", value = "Religion:")
	@Localize(locale = "ru", value = "Религия:")
	@Localize(locale = "es", value = "Religión:")
	private static String	RELIGION;

	static {
		Localization.initialize();
	}

	/**
	 * Creates a new identity panel.
	 *
	 * @param sheet The sheet to display the data for.
	 */
	public IdentityPanel(CharacterSheet sheet) {
		super(new ColumnLayout(2, 2, 0), IDENTITY);
		createLabelAndField(this, sheet, Profile.ID_NAME, NAME, null, SwingConstants.LEFT);
		createLabelAndField(this, sheet, Profile.ID_TITLE, TITLE, null, SwingConstants.LEFT);
		createLabelAndField(this, sheet, Profile.ID_RELIGION, RELIGION, null, SwingConstants.LEFT);
	}
}

File: ux/identity_panel.go
// Copyright (c) 1998-2025 by Richard A. Wilkes. All rights reserved.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, version 2.0. If a copy of the MPL was not distributed with
// this file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// This Source Code Form is "Incompatible With Secondary Licenses", as
// defined by the Mozilla Public License, version 2.0.

package ux

import (
	"github.com/richardwilkes/gcs/v5/model/colors"
	"github.com/richardwilkes/gcs/v5/model/gurps"
	"github.com/richardwilkes/toolbox/i18n"
	"github.com/richardwilkes/unison"
	"github.com/richardwilkes/unison/enums/align"
)

const (
	identityPanelFieldPrefix     = "identity:"
	identityPanelNameFieldRefKey = identityPanelFieldPrefix + "name"
)

// IdentityPanel holds the contents of the identity block on the sheet.
type IdentityPanel struct {
	unison.Panel
	entity    *gurps.Entity
	targetMgr *TargetMgr
	prefix    string
}

// NewIdentityPanel creates a new identity panel.
func NewIdentityPanel(entity *gurps.Entity, targetMgr *TargetMgr) *IdentityPanel {
	p := &IdentityPanel{
		entity:    entity,
		targetMgr: targetMgr,
		prefix:    identityPanelFieldPrefix,
	}
	p.Self = p
	p.SetLayout(&unison.FlexLayout{
		Columns:  2,
		HSpacing: 4,
	})
	p.SetLayoutData(&unison.FlexLayoutData{
		HAlign: align.Fill,
		VAlign: align.Fill,
		HGrab:  true,
	})
	p.SetBorder(unison.NewCompoundBorder(&TitledBorder{Title: i18n.Text("Identity")}, unison.NewEmptyBorder(unison.Insets{
		Top:    1,
		Left:   2,
		Bottom: 1,
		Right:  2,
	})))
	p.DrawCallback = func(gc *unison.Canvas, rect unison.Rect) { drawBandedBackground(p, gc, rect, 0, 2, nil) }
	InstallTintFunc(p, colors.TintIdentity)

	title := i18n.Text("Name")
	nameField := NewStringPageField(p.targetMgr, identityPanelNameFieldRefKey, title,
		func() string { return p.entity.Profile.Name },
		func(s string) { p.entity.Profile.Name = s })
	p.AddChild(NewPageLabelWithRandomizer(title,
		i18n.Text("Randomize the name using the current ancestry"), func() {
			p.entity.Profile.Name = p.entity.Ancestry().RandomName(
				gurps.AvailableNameGenerators(gurps.GlobalSettings().Libraries()), p.entity.Profile.Gender)
			SetTextAndMarkModified(nameField.Field, p.entity.Profile.Name)
		}))
	nameField.ClientData()[SkipDeepSync] = true
	p.AddChild(nameField)

	title = i18n.Text("Title")
	p.AddChild(NewPageLabelEnd(title))
	titleField := NewStringPageField(p.targetMgr, p.prefix+"title", title,
		func() string { return p.entity.Profile.Title },
		func(s string) { p.entity.Profile.Title = s })
	titleField.ClientData()[SkipDeepSync] = true
	p.AddChild(titleField)

	title = i18n.Text("Organization")
	p.AddChild(NewPageLabelEnd(title))
	orgField := NewStringPageField(p.targetMgr, p.prefix+"org", title,
		func() string { return p.entity.Profile.Organization },
		func(s string) { p.entity.Profile.Organization = s })
	orgField.ClientData()[SkipDeepSync] = true
	p.AddChild(orgField)
	return p
}
